File: ios/CylinderSeal/ViewModels/BusinessOnboardingViewModel.swift
import Foundation

enum BusinessKind: String, CaseIterable, Identifiable {
    case pos = "business_pos"
    case electronic = "business_electronic"

    var id: String { rawValue }

    var label: String {
        switch self {
        case .pos: return "Physical point of sale"
        case .electronic: return "Electronic / online commerce"
        }
    }

    var subtitle: String {
        switch self {
        case .pos: return "Shops, market stalls, services with a physical till."
        case .electronic: return "E-commerce, B2B, SaaS. Includes API key access."
        }
    }
}

@MainActor
final class BusinessOnboardingViewModel: ObservableObject {
    @Published var kind: BusinessKind = .pos
    @Published var legalName = ""
    @Published var commercialRegistrationId = ""
    @Published var taxId = ""
    @Published var industryCode = ""
    @Published var registeredAddress = ""
    @Published var contactEmail = ""
    @Published var submitted = false
    @Published var statusMessage: String?
    @Published var errorMessage: String?
    @Published var busy = false

    private let wallet: WalletKeyManager
    private let api: BusinessApiClient

    init(wallet: WalletKeyManager, api: BusinessApiClient) {
        self.wallet = wallet
        self.api = api
    }

    func submit() {
        if [legalName, commercialRegistrationId, taxId, industryCode, registeredAddress, contactEmail]
            .contains(where: { $0.trimmingCharacters(in: .whitespaces).isEmpty })
        {
            errorMessage = "All fields are required"
            return
        }

        Task {
            busy = true
            errorMessage = nil
            defer { busy = false }
            do {
                let publicKey = try wallet.loadPublicKey()
                let userId = try MobileCore.userIdFromPublicKey(publicKey)
                let req = BusinessApiClient.RegisterRequest(
                    user_id: userId,
                    account_type: kind.rawValue,
                    legal_name: legalName,
                    commercial_registration_id: commercialRegistrationId,
                    tax_id: taxId,
                    industry_code: industryCode,
                    registered_address: registeredAddress,
                    contact_email: contactEmail,
                    authorized_signer_public_keys_hex: [publicKey.toHex()]
                )
                let resp = try await api.register(req)
                submitted = true
                statusMessage = "Registered: \(resp.status). CBI ops will verify your commercial registration before activation."
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }
}

File: android/feature/feature-business/src/main/kotlin/com/modernecotech/cylinderseal/feature/business/BusinessOnboardingViewModel.kt
package com.modernecotech.cylinderseal.feature.business

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.modernecotech.cylinderseal.core.common.toHex
import com.modernecotech.cylinderseal.core.cryptography.WalletKeyManager
import com.modernecotech.cylinderseal.core.ffi.MobileCore
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

enum class BusinessKind { POS, ELECTRONIC }

data class BusinessOnboardingState(
    val kind: BusinessKind = BusinessKind.POS,
    val legalName: String = "",
    val commercialRegistrationId: String = "",
    val taxId: String = "",
    val industryCode: String = "",
    val registeredAddress: String = "",
    val contactEmail: String = "",
    val submitted: Boolean = false,
    val statusMessage: String? = null,
    val errorMessage: String? = null,
    val busy: Boolean = false,
)

@HiltViewModel
class BusinessOnboardingViewModel @Inject constructor(
    private val wallet: WalletKeyManager,
    private val api: BusinessApi,
) : ViewModel() {
    private val _state = MutableStateFlow(BusinessOnboardingState())
    val state: StateFlow<BusinessOnboardingState> = _state

    fun setKind(v: BusinessKind) = update { it.copy(kind = v) }
    fun setLegalName(v: String) = update { it.copy(legalName = v) }
    fun setRegistration(v: String) = update { it.copy(commercialRegistrationId = v) }
    fun setTaxId(v: String) = update { it.copy(taxId = v) }
    fun setIndustry(v: String) = update { it.copy(industryCode = v.filter(Char::isDigit).take(4)) }
    fun setAddress(v: String) = update { it.copy(registeredAddress = v) }
    fun setEmail(v: String) = update { it.copy(contactEmail = v.trim()) }

    fun submit() {
        val s = _state.value
        if (s.legalName.isBlank() || s.commercialRegistrationId.isBlank() ||
            s.taxId.isBlank() || s.industryCode.isBlank() ||
            s.registeredAddress.isBlank() || s.contactEmail.isBlank()
        ) {
            update { it.copy(errorMessage = "All fields are required") }
            return
        }

        viewModelScope.launch {
            update { it.copy(busy = true, errorMessage = null) }
            try {
                val publicKey = wallet.loadPublicKey()
                val userId = MobileCore.userIdFromPublicKey(publicKey)
                val accountType = when (s.kind) {
                    BusinessKind.POS -> "business_pos"
                    BusinessKind.ELECTRONIC -> "business_electronic"
                }
                val resp = api.register(
                    RegisterRequest(
                        user_id = userId,
                        account_type = accountType,
                        legal_name = s.legalName,
                        commercial_registration_id = s.commercialRegistrationId,
                        tax_id = s.taxId,
                        industry_code = s.industryCode,
                        registered_address = s.registeredAddress,
                        contact_email = s.contactEmail,
                        authorized_signer_public_keys_hex = listOf(publicKey.toHex()),
                    )
                )
                update {
                    it.copy(
                        busy = false,
                        submitted = true,
                        statusMessage = "Registered: ${resp.status}. CBI ops will verify your commercial registration before activation.",
                    )
                }
            } catch (t: Throwable) {
                update {
                    it.copy(busy = false, errorMessage = t.message ?: "Registration failed")
                }
            }
        }
    }

    private inline fun update(transform: (BusinessOnboardingState) -> BusinessOnboardingState) {
        _state.value = transform(_state.value)
    }
}
